default/python/universe_generation/galaxy.py

```diff
@@ -1,35 +1,349 @@
+
+    def complete_sets(self):
+        """ returns a list of list of all sets O(n)."""
+        ret = defaultdict(list)
+        for pos in self.dsets.keys():
+            ret[self.root(pos)].append(pos)
+        return ret.values()
+
+
+class Clusterer(object):
+    """
+    Computes all clusters of positions separated by more than MAX_STARLANE_LENGTH
+
+    This creates a disjoint set of all of the positions
+    Adds all of the seperations less than MAX_STARLANE_LENGTH
+    Then returns the clusters of stars when requested
+    """
+    def __init__(self, positions, adjacency_grid):
+        self.adjacency_grid = adjacency_grid
+        self.dsets = DisjointSets()
+
+        for pos in positions:
+            self._add(pos)
+
+        self.clusters = set()
+        for cluster in self.dsets.complete_sets():
+            # Decorate each cluster with its length to speed sorting
+            self.clusters.add((len(cluster), frozenset(cluster)))
+
+    def __len__(self):
+        return len(self.clusters)
+
+    def _add(self, pos):
+        """Adds pos."""
+        for neighbor in self.adjacency_grid.neighbors(pos):
+            # print "Clusterer adding ", pos,  " and ", neighbor
+            self.dsets.link(pos, neighbor)
+
+    def smallest_isolated_cluster(self):
+        """
+        Return None if all of the positions are in one cluster
+        or the smallest cluster
+        """
+        # Sort (min) and return the undecorated cluster.
+        return None if len(self.clusters) < 2 else min(self.clusters)[1]
+
+    def stitch_clusters(self, p1, p2, stitches):
+        """
+        Stitch the clusters containing ''p1'' and ''p2''
+        together with the positions in ''stitches''
+
+        This assumes that the stitching is correctly formed.
+        ''p1'' and ''p2'' are the closest positions between
+        the clusters and the positions in ''stitches'' are only
+        between ''p1'' and ''p2''
+        """
+        len_dset1 = None
+        len_dset2 = None
+        for len_dset in self.clusters:
+            if p1 in len_dset[1]:
+                len_dset1 = len_dset
+            if p2 in len_dset[1]:
+                len_dset2 = len_dset
+        assert len_dset1 and len_dset2, "p1 and p2 must be points in disjoint sets of positions"
+        self.clusters.remove(len_dset1)
+        self.clusters.remove(len_dset2)
+        # Decorate the new cluster with its length to speed sorting
+        new_set = len_dset1[1].union(len_dset2[1]).union(frozenset(stitches))
+        self.clusters.add((len(new_set), new_set))
+
+
+def stitching_positions(p1, p2):
+    """
+    Returns a list of positions between p1 and p2 between MIN_SYSTEM_SEPARATION and MAX_STARLANE_LENGTH apart
+    """
+    assert 2 * universe_tables.MIN_SYSTEM_SEPARATION < universe_tables.MAX_STARLANE_LENGTH,\
+        "MAX_STARLANE_LENGTH must be twice MIN_SYSTEM_SEPARATION to allow extra positions to \
+        be added to enforce MAX_STARLANE_LENGTH"
```

Comment:
`MAX_STARLANE_LENGTH` and `MIN_SYSTEM_SEPARATION` are supposed to be parameters that can be customized by players, so failing here with an assertion isn't a good idea. General policy is to abort execution only if it can't be avoided at all, otherwise log an error and try to continue.

In case of the other assertion check above (`AdjacencyGrid` init) failing hard probably can't be avoided, because if that condition isn't met, star system positioning is screwed. But here you can try to continue, just abort the process of stitching together the disjointed clusters. Log an error using the `report_error` function of the `util` module, which will ensure the player gets a proper warning at the start of the game.